src/wallet/wallet2.cpp

```diff
@@ -5538,6 +5785,100 @@ size_t wallet2::import_outputs(const std::vector<tools::wallet2::transfer_detail
   return m_transfers.size();
 }
 //----------------------------------------------------------------------------------------------------
+rct::key wallet2::get_multisig_k(size_t idx) const
+{
+  CHECK_AND_ASSERT_THROW_MES(m_multisig, "Wallet is not multisig");
+  CHECK_AND_ASSERT_THROW_MES(idx < m_transfers.size(), "Index out of range");
+  THROW_WALLET_EXCEPTION_IF(m_transfers[idx].m_multisig_k == rct::zero(), error::multisig_export_needed);
+  return m_transfers[idx].m_multisig_k;
+}
+//----------------------------------------------------------------------------------------------------
+rct::multisig_kLRki wallet2::get_multisig_LRki(size_t n, const rct::key &k) const
+{
+  rct::multisig_kLRki kLRki;
+  kLRki.k = k;
+  rct::scalarmultBase(kLRki.L, kLRki.k);
+  crypto::generate_key_image(m_transfers[n].get_public_key(), rct::rct2sk(kLRki.k), (crypto::key_image&)kLRki.R);
+  kLRki.ki = rct::ki2rct(m_transfers[n].m_key_image);
+  return kLRki;
+}
+//----------------------------------------------------------------------------------------------------
+rct::multisig_kLRki wallet2::get_multisig_composite_LRki(size_t n, const rct::key &k) const
+{
+  rct::multisig_kLRki kLRki = get_multisig_LRki(n, k);
+  const transfer_details &td = m_transfers[n];
+  crypto::public_key tx_key = get_tx_pub_key_from_received_outs(td);
+  cryptonote::keypair in_ephemeral;
+  wallet_generate_key_image_helper(get_account().get_keys(), tx_key, td.m_internal_output_index, in_ephemeral, (crypto::key_image&)kLRki.ki);
+  for (const auto &info: td.m_multisig_info)
+  {
+    rct::addKeys(kLRki.ki, kLRki.ki, rct::ki2rct(info.m_partial_key_image));
+    rct::addKeys(kLRki.L, kLRki.L, info.m_L);
+    rct::addKeys(kLRki.R, kLRki.R, info.m_R);
+  }
+  return kLRki;
+}
+//----------------------------------------------------------------------------------------------------
+std::vector<tools::wallet2::multisig_info> wallet2::export_multisig()
+{
+  std::vector<tools::wallet2::multisig_info> info;
+
+  info.resize(m_transfers.size());
+  for (size_t n = 0; n < m_transfers.size(); ++n)
+  {
+    transfer_details &td = m_transfers[n];
+    crypto::public_key tx_key = get_tx_pub_key_from_received_outs(td);
+    cryptonote::keypair in_ephemeral;
+    crypto::key_image ki;
+    td.m_multisig_k = rct::skGen();
```

Comment:
Isn't this rarely needed, because `td.m_multisig_k` is already set inside `process_new_transaction()`? I think the only situation to have to set this field again would be when a complete multisig tx was sent to the daemon (which means `td.m_multisig_k` was set to zero in `commit_tx()`) but was never mined and dropped from the mempool for some reason. To take that into account, I'd do as follows:

    if (td.m_multisig_k == rct::zero())
      td.m_multisig_k = rct::skGen();